drivers/sony/displays/bravia.cr

```diff
@@ -0,0 +1,226 @@
+    "Mirror" => "5000",
+    "Vga"    => "6000",
+  }
+
+  CONVERT_LOOKUP = INPUTS_CONVERT.invert
+
+  include Interface::InputSelection(Inputs)
+
+  def switch_to(input : Inputs)
+    parsed_input = input.to_s.scan(/[^0-9]+|\d+/)
+    index = parsed_input.size < 2 ? "1" : parsed_input[1][0]
+    input_joined = INPUTS_CONVERT[parsed_input[0][0]] + index.rjust(5, '0')
+    request(:input, input_joined)
+    self[:input] = "#{input}"
+    input?
+  end
+
+  def input?
+    query(:input, priority: 0)
+  end
+
+  def on_load
+    self[:volume_min] = 0
+    self[:volume_max] = 100
+  end
+
+  def connected
+    schedule.every(30.seconds, true) do
+      do_poll
+    end
+  end
+
+  def disconnected
+    schedule.clear
+  end
+
+  def power(state : Bool)
+    request(:power, state)
+    logger.debug { "Sony display requested power #{state ? "on" : "off"}" }
+    power?
+  end
+
+  def power?
+    query(:power)
+  end
+
+  def mute(
+    state : Bool = true,
+    index : Int32 | String = 0,
+    layer : MuteLayer = MuteLayer::AudioVideo
+  )
+    request(:mute, state)
+    mute?
+  end
+
+  def unmute
+    mute false
+  end
+
+  def mute?
+    query(:mute, priority: 0)
+  end
+
+  def mute_audio(state : Bool = true)
+    request(:audio_mute, state)
+    audio_mute?
+  end
+
+  def unmute_audio
+    mute_audio false
+  end
+
+  def audio_mute?
+    query(:audio_mute, priority: 0)
+  end
+
+  def volume(level : Int32)
+    request(:volume, level.to_i)
+    volume?
+  end
+
+  def volume?
+    query(:volume, priority: 0)
+  end
+
+  def do_poll
+    if self[:power]?
+      input?
+      mute?
+      audio_mute?
+      volume?
+    end
+  end
+
+  enum MessageType : UInt8
+    Answer  = 0x41
+    Control = 0x43
+    Enquiry = 0x45
+    Notify  = 0x4e
+
+    def control_character
+      case self
+      in Answer  then "A"
+      in Control then "C"
+      in Enquiry then "E"
+      in Notify  then "N"
+      end
+    end
+  end
+
+  def received(data, task)
+    parsed_data = convert_binary(data[3..6])
+    cmd = RESPONSES[parsed_data]
+    param = data[7..-1]
+
+    return task.try(&.abort("error")) if param.first? == ERROR
+    case MessageType.from_value?(data[2])
+    when MessageType::Answer
+      update_status cmd, param
+      task.try &.success
+    when MessageType::Notify
+      update_status cmd, param
+    else
+      logger.debug { "Unhandled device response: #{data[2]}" }
+      task.try &.abort("Unhandled device response")
+    end
+  end
+
+  COMMANDS = {
+    ir_code:           "IRCC",
+    power:             "POWR",
+    volume:            "VOLU",
+    audio_mute:        "AMUT",
+    mute:              "PMUT",
+    channel:           "CHNN",
+    tv_input:          "ISRC",
+    input:             "INPT",
+    toggle_mute:       "TPMU",
+    pip:               "PIPI",
+    toggle_pip:        "TPIP",
+    position_pip:      "TPPP",
+    broadcast_address: "BADR",
+    mac_address:       "MADR",
+  }
+  RESPONSES = COMMANDS.to_h.invert
```

Comment:
Could be cleaner if contained in an enum. That would allow for total matching on the `case` statement down on 209-223.
Pragmatically, however, I think this is in a good place.